README.md
This project has been automatically exported from code.google.com/p/prs-plus
It seems to be no longer under active development.
Many thanks to all contributors who made this project possible.

Owners of supported devices can find the code, issues, installer, uninstaller and help files here.
In difference to other mirrors here on github, I try to bring the files in order.
You can download individual installers from prs-plus/downloads and read the user guides and help files in the wiki.



Changelog.md
### 2.0.18 ###
  * ALL: Added scaling option to Standby Image
  * x50: Updated translations

### 2.0.17 release candidate 3 (2011.12.12) ###
  * x50: Fixed [Issue #244](https://code.google.com/p/prs-plus/issues/detail?id=#244) "books deleted using PC do not dissapear if scanning is "disabled (load cache)""
  * ALL: Fixed missing translation
  * 600: Fixed [Issue #237](https://code.google.com/p/prs-plus/issues/detail?id=#237) Menu Customization is buggy
  * ALL: Fixed [Issue #234](https://code.google.com/p/prs-plus/issues/detail?id=#234) "Mah Jong reports "no more moves" when moves are still available"


### 2.0.16 release candidate 2 (2011.11.26) ###
  * ALL: Minor translation fixes (RUS, TUR)
  * ALL: StandbyImage: Fix for scaling bug in Sony code
  * 300/505/600: Added $a0 (no wrap space) to fonts (to be added to x50 in 2.1.x)

### 2.0.15 release candidate (2011.11.21) ###
  * ALL: Fixed #171 "The "Copy to IM..." menu items are not present in Card via Mount, if card scanning is not disabled"
  * ALL: Fixed bug that prevented SD/MS card scan mode from being changed on the fly
  * ALL: Fixed #214 MSG\_COPYING\_BOOK not translated
  * ALL: EPUB Css styles are now sorted
  * ALL: Fixes #215 fb2epub converter doesn't work with cards with disabled scanning
  * 350/650: Improved Dutch translation (drMerry)
  * 350/650: Fixed "Continue searching from the begining doesn't work" (Sony bug)
  * 350/650: Added 1-Column Split (quisvir)
  * x50: "Cycle booklist" action now also cycles through collections (quisvir)
  * x50: Added option to treat periodicals as books (quisvir)
  * x50: Fixed #211  that caused SD/MS card scan options to be ignored on the first boot
  * 600/x50: Added Page Turn by Single Tap (quisvir)
  * 600/x50: Close reading popup menu (dict etc) and cancel selection by tapping page (quisvir)
  * 600/x50: Fixed #207 Collection sorting broken for Cyrillic
  * 600: Fixed #197 "SD/MS card via mount" doesn't handle non-latin characters
  * 600: Added keyCodes and Hold keyCodes to key bindings
  * 600: Fonts updated to the latest version
  * 600: Added Turkish translation
  * 600: Changed "holding option button" default action to do "go to parent view"
  * 600: Shifted games to the bottom of "More" list

### 2.0.14 beta (2011.09.22) ###
  * x50: Fixed #128 books/collections sorting
  * x50: Fixed "Coming back to Home from an other menu takes long"
  * 300/500: Fixed CoverPage as StandbyImage, fixed problem with next page after sleepmode
  * ALL: BrowseFolder patched with Shura1oplots FileSize in Comment & added FileType
  * ALL: Fixed russian localisation of "No Book, 1 Book, x Books"

### 2.0.13 beta (2011.09.14) ###
  * 505: Fixed #126 "images/audio lists are empty"
  * 505: Fixed #139 "Menu customizer values are not translated"
  * 505: Fixed #150 "Some options in the setup menu are not translated" needs reflash! not working with beta-pack on SD-Card
  * 505: Added #39 "Hold joypad arrows events"
  * x50: Fixed #120 "No keyboard in SP-EN dictionary"
  * x50: Enabled "half-page-mask" in landscape mode (like PRS-505), added option to set PDF-border-color to white
  * x50: Added option to preset custom Contrast and Brightness values (quisvir)
  * x50: Added option to toggle NEW-flag manually, hide standard collections and show reading-progress in home menu and thumbnail views (quisvir)
  * x50: Added Home Menu Booklist customization (quisvir)
  * 600/x50: Games moved into Games node
  * 600/x50: Added various Touch-Screen related options (disable dictionary and page-turn-gestures)
  * 600/x50: Enabled panning while ZoomLocked (quisvir)
  * ALL: Fonts: Added Korean alphabet (3131-314E unicode range), updated Czech d t l L letters, added up/down arrows
  * ALL: Added #70 "Make node containers accessible via menu customizer"
  * ALL: Added #24 "Displaying first page of the book on standby"
  * ALL: Screenshot: captured image is now  visible to the user, not only to the system :)
  * ALL: BrowseFolders:  added optional ".." item (by Shura1oplot)
  * ALL: Action are now grouped and have icons (by Shura1oplot)
  * ALL: New icons for utils and games (by surquizu, Shura1oplot, ?)
  * ALL: Updated existing games (Chess now comes with puzzles)
  * ALL: Added MineSweeper (by Mark Nord), Draughts, XO-Cubed and Solitaire (by Ben Chenoweth)
  * ALL: Added Calendar app (by Ben Chenoweth)
  * All: Added various "Standby" options (display book cover, last screen) (by Mark Nord)

### 2.0.12 beta (2011.05.26) ###
  * 505: Fixed German translation
  * ALL: Fixed Courier font name
  * ALL: ~~Sudoku can be exited by pressing home button~~ (not in 950, maybe other models)
  * 350/650: Fixed custom CSS path

### 2.0.11 beta ###
  * 300: Fixed broken XML in Spanish translation
  * 300: Added Ukrainian localization by Bookoman
  * 300/505: Fixed "actions do not work"
  * 300/505: Added French localization by sengian (installer)
  * 300/505: Fixed bug related to "Skip book menu" option
  * 505: Restored localizations
  * 505: Restored 1.1.3 menu layout with "Multimedia"
  * 600: Added Portuguese localization by OTNeto
  * 600: Added Czech localization by milanv
  * 600/x50: Fixed keyboard: "aaaa" is shown instead of ascented (popup) letters
  * 600/x50: TextScale addon was included
  * x50: Reduced statusbar index / clock font size, moved "playing" indicator to the left
  * x50: Fixed "Periodicals"
  * x50: Fixed sorting in "books" for non latin alphabets
  * 950: Added GMT+10 timezone
  * ALL: Screenshot: captured image is now immediatelly visible to the system
  * ALL: Folders: Added option to disable scanning without loading cache (as it was in 1.1.3)
  * ALL: Folders: Added "via mount" card access (uses Linux "mount" command instead of mtools like buggy utility)
  * ALL: Added "LRF Text Scale" addon that allows to customize zoom levels

### 2.0.10 alpha ###
  * x50: Fixed doRotate action
  * x50: Fixed #68 x50: Deleting books opened via Book History is bugged

script/core/core_ui.js
// Name: UI
// Description: User interface related methods and constants
// Author: kartu
//
// History:
//	2010-03-14 kartu - Initial version, refactored from Utils
//	2010-04-05 kartu - Removed stale code, added logging to getValue
//	2010-04-10 kartu - Improved error reporting
//	2010-04-17 kartu - Removed global var
//	2010-04-25 kartu - Marked setLevel Core.ui.ContainerNode as constructor
//	2010-04-27 kravitz - Added showMsg()
//	2010-04-27 kravitz - Fixed getValue() and getKind()
//	2010-06-29 kartu - Adapted for 300
//				Added string parameter support to showMsg
//	2010-08-23 kartu - Added doBlink method
//	2010-11-10 kartu - Fixed homekind field name (was homeKind)
//				Added "shortTitle" field support (text to fit on small buttons of touch readers etc)
//	2011-01-31 kartu - Added multiPage field (x50) to enable scrollbars
//	2011-02-06 kartu - Added Core.ui.updateScreen
//	2011-03-03 kartu - Added "duration" option to showMsg
//	2011-03-19 kartu - Changed showMsg, duration is now in seconds, changed "sleep" mechanism, shell command is used
//	2011-04-24 kartu - Changed showMsg to honor EOLs 
//	2011-06-18 kartu - Fixed "update() wasn't called by container mode"
//	2011-06-26 kartu - Again fixed "update() wasn't called by container mode"
//	2011-10-01 quisvir - Added guideArea property for containers
//	2011-12-29 Ben Chenoweth - Added setCurrentNode(node)
//	2012-01-29 Mark Nord - added "infinite" duration to Core.ui.showMsg

try {
	var doSetNodeIcon = function (node, icon) {
		if (typeof icon === "number") {
			node.kind = icon;
		} else {
			node.kind = Core.config.compat.NodeKinds.getIcon(icon);
			node.homekind = Core.config.compat.NodeKinds.getIcon(icon, "home");
			node.homelargekind = Core.config.compat.NodeKinds.getIcon(icon, "homeLarge");				
		}
	};


	// Fix for "update()" otherwise code fails not finding "children" field etc
	var containerUpdate = function (model) {
		var i, n, nodes, node;
		nodes = this.nodes;
		if (nodes) {
			for (i = 0, n = nodes.length; i < n; i++) {
				try {
					node = nodes[i];
					node.update(model);
				} catch (e) {
					log.error("calling update for node: " + node);
				}
			}
		}
	};
	
	var doCreateContainerNode = function (arg, prototype) {
		var obj = xs.newInstanceOf(prototype);
		obj.onEnter = "onEnterDefault";
		obj.onSelect = "onSelectDefault";
		
		if (typeof arg !== "undefined") {
			if (arg.hasOwnProperty("parent")) {obj.parent = arg.parent;}
			if (arg.hasOwnProperty("title")) {obj.title = arg.title;}
			if (arg.hasOwnProperty("shortName")) {obj.shortName = arg.shortName;}
			if (arg.hasOwnProperty("name")) {
				obj.name = arg.name;
			} else {
				obj.name = arg.title;
			}
			if (arg.hasOwnProperty("comment") && (typeof arg.comment !== "undefined")) {
				if (typeof arg.comment === "function") {
					obj._mycomment = arg.comment;
				} else {
					obj.comment = arg.comment;
				}
			} else {
				obj._mycomment = "";
			}
			doSetNodeIcon(obj, arg.icon);
			if (arg.hasOwnProperty("guideArea")) {obj.guideArea = arg.guideArea;}
			if (arg.hasOwnProperty("separator")) {obj.separator = arg.separator;}
			if (arg.hasOwnProperty("construct")) {obj.construct = arg.construct;}
			if (arg.hasOwnProperty("destruct")) {obj.destruct = arg.destruct;}
		}
		obj.nodes = [];
		obj.multiPage = true;

		return obj;
	};

	Core.ui = {
		// Creates "container" node, that displayes nodes in this.nodes[] array
		// Arguments:
		//	arg, can have the following fields:
		//		parent - parent node
		//		title - title of this node (shown on top of the screen, when inside the node)
		//		name - name of this node (shown in lists, if none supplied, title is used instead)
		//		comment - comment text (shown on the right bottom in list mode)
		//		icon - either string (a key) or value, determines which icon to show (see Core.config.compat.NodeKinds)
		//		separator - if equals to 1, node's bottom line will be shown in bold
		//		constructor - method to be called to populate child node list
		//		destructor - method to be called, when child node list can be destroyed
		//		
		createContainerNode: function (arg) {
			var node;
			if (arg.hasOwnProperty("construct") || arg.hasOwnProperty("destruct")) {
				node = doCreateContainerNode(arg, FskCache.tree.xdbNode);
			} else {
				node = doCreateContainerNode(arg, FskCache.tree.containerNode);
				node.update = containerUpdate;
			}
			return node;
		}
	};

	Core.ui.setNodeIcon = function (node, icon) {
		return doSetNodeIcon(node, icon);
	};

	// Forces screen update
	//
	Core.ui.updateScreen = function () {
		FskUI.Window.update.call(kbook.model.container.getWindow());
	};

	// Shows "msgs" for given amount of time
	// Note: duration is rather unreliable, and message might hang on the screen until user presses button / touches screen
	//
	// Arguments:
	//	msgs - array of strings
	//	duration - for how long to show in seconds, 0 or negative number means don't wipe. 2 is the default value.
	//
	Core.ui.showMsg = function (msgs, duration) {
		var cnt, win, gap, spc, ms_w, ms_h, i, b, w, h, x, y, x1, y1,
			oldTextStyle, oldTextSize, oldPenColor;

		
		if (typeof msgs === "string") {
			msgs = msgs.split("\n");
		}
		cnt = msgs.length;
		if (cnt === undefined || cnt === 0) {
			return;
		}
		win = kbook.model.container.getWindow();
		
		// Save old styles
		oldTextStyle = win.getTextStyle();
		oldTextSize = win.getTextSize();
		oldPenColor = win.getPenColor();
		
		// Settings
		gap = 20;
		spc = 10;
		win.setTextStyle(1);
		win.setTextSize(22);
		// Calc
		ms_w = [];
		ms_h = [];
		for (i = 0; i < cnt; i++) {
			b = win.getTextBounds(msgs[i]);
			ms_w.push(b.width);
			ms_h.push(b.height);
		}
		w = Math.max.apply( Math, ms_w ) + gap * 2;
		h = ms_h[0] * cnt + spc * (cnt - 1) + gap * 2;
		b = win.getBounds();
		x = Math.max(0, (b.width - w) / 2);
		y = Math.max(0, (b.height - h) / 2);
		// Drawing
		win.beginDrawing();
		win.setPenColor(Color.white);
		win.fillRectangle(x, y, w, h);
		win.setPenColor(Color.black);
		win.frameRectangle(x, y, w, h);
		win.frameRectangle(x + 1, y + 1, w - 2, h - 2);
		x1 = x + gap;
		y1 = y + gap;
		for (i = 0; i < cnt; i++) {
			win.drawText(msgs[i], x1, y1, ms_w[i], ms_h[i]);
			y1 += ms_h[i] + spc;
		}
		win.endDrawing();
		
		// Restore pen color, text size & style
		win.setTextStyle(oldTextStyle);
		win.setTextSize(oldTextSize);
		win.setPenColor(oldPenColor);
		
		// Pause
		if (duration === undefined) {
			duration = 2;
		}
		if (duration === "infinite") { return }
		if (duration > 0) {
			try {
				Core.shell.exec("sleep " + duration);
				if (kbook.model.currentNode && kbook.model.currentNode.name === "About") {
					// Crashes in about screen on 650,? if only part of it is invalidated FIXME: figure why
					win.invalidate();
				} else {
					win.invalidate(x, y, w, h);
				}
			} catch (ignore) {
			}
		}
	};
	
	/**
	* Blinks with Sony's standard "error" icon
	*/
	Core.ui.doBlink = function () {
		kbook.model.doBlink();
	};
	
	/**
	* @returns currently focused node
	*/
	Core.ui.getCurrentNode = function () {
		if (kbook.model.current) {
			// 300/505
			return kbook.model.current;
		} else {
			// 600+
			return kbook.model.currentNode;
		}
	};

	/**
	* @changes the currently focused node
	*/
	Core.ui.setCurrentNode = function (node) {
		if (kbook.model.current) {
			// 300/505
			kbook.model.current = node;
		} else {
			// 600+
			kbook.model.currentNode = node;
		}
	};

	// Little hack to allow easy changing of node title, comment etc
	// FIXME try to get rid of this
	var oldGetValue = Fskin.tableData.getValue;
	Fskin.tableData.getValue = function (node, field) {
		try {
			var myVal = node["_my" + field];
			if (myVal !== undefined) {
				if (typeof myVal === "function") {
					return myVal.call(node);
				}
				return myVal;
			}
		} catch (e) {
			log.error("in _my getValue: field '" + field + "' node '" + node.name + "': " + e);
		}
		try {
			return oldGetValue.apply(this, arguments);
		} catch (e2) {
			log.error("in getValue: " + e2);
			return "error: " + e2;
		}
	};
} catch (e) {
	log.error("initializing core-ui", e);
}
